luigi/contrib/dropbox.py

```diff
@@ -0,0 +1,275 @@
+# the License.
+#
+
+from __future__ import absolute_import
+
+import datetime
+import logging
+import ntpath
+import os
+import random
+import tempfile
+from contextlib import contextmanager
+
+import luigi.format
+from luigi.target import FileSystem, FileSystemTarget, AtomicLocalFile
+
+logger = logging.getLogger('luigi-interface')
+
+try:
+    import dropbox.dropbox
+    import dropbox.exceptions
+    import dropbox.files
+except ImportError:
+    logger.warning("Loading Dropbox module without the python package dropbox (https://pypi.org/project/dropbox/). "
+                   "Will crash at runtime if Dropbox functionality is used.")
+
+
+class DropboxClient(FileSystem):
+    """
+    Dropbox client for authentication, designed to be used by the :py:class:`DropboxTarget` class.
+    """
+
+    def __init__(self, token):
+        """
+        :param str token: Dropbox Oauth2 Token. See :class:`DropboxTarget` for more information about generating a token
+        """
+        if not token:
+            raise ValueError("The token parameter must contain a valid Dropbox Oauth2 Token")
+
+        try:
+            conn = dropbox.dropbox.Dropbox(oauth2_access_token=token, user_agent="Luigi")
+        except Exception as e:
+            raise Exception("Cannot connect to Dropbox. Check your Internet connection and the token. \n" + repr(e))
+
+        self.token = token
+        self.conn = conn
+
+    def exists(self, path):
+        if path == '/':
+            return True
+        try:
+            md = self.conn.files_get_metadata(path)
+            return bool(md)
+        except dropbox.exceptions.ApiError as e:
+            if isinstance(e.error.get_path(), dropbox.files.LookupError):
+                return False
+            else:
+                raise e
+
+    def remove(self, path, recursive=True, skip_trash=True):
+        if not self.exists(path):
+            return False
+        self.conn.files_delete_v2(path)
+        return True
+
+    def mkdir(self, path, parents=True, raise_if_exists=False):
+        if self.exists(path):
+            if not self.isdir(path):
+                raise luigi.target.NotADirectory()
+            elif raise_if_exists:
+                raise luigi.target.FileAlreadyExists()
+            else:
+                return
+
+        self.conn.files_create_folder_v2(path)
+
+    def isdir(self, path):
+        if path == '/':
+            return True
+        try:
+            md = self.conn.files_get_metadata(path)
+            return isinstance(md, dropbox.files.FolderMetadata)
+        except dropbox.exceptions.ApiError as e:
+            if isinstance(e.error.get_path(), dropbox.files.LookupError):
+                return False
+            else:
+                raise e
+
+    def listdir(self, path, **kwargs):
+        dirs = []
+        lister = self.conn.files_list_folder(path, recursive=True, **kwargs)
+        dirs.extend(lister.entries)
```

Comment:
why can't these 3 lines just become `dirs = self.conn.files_list_folder(...)`?

Reply:
We need 'lister' because we will use the `lister.entries` (with the list of directories) and also `lister.has_more` and `lister.cursor`.